### PR TITLE
State fixed-size array: `block_roots` and `state_roots`

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Refactoring

**What does this PR do? Why is it needed?**

Replaces proto definition of `block_roots` and `state_roots` in the state from `[][]byte` to `[8192][32]byte`. 

**Which issues(s) does this PR fix?**

Part of #9820 

**Other notes for review**

`go build ./...` is successful.

### CODE REVIEW

File: beacon-chain/state/custom-types/state_arrays.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ var _ fssz.Marshaler = (*StateRoots)(nil)
 var _ fssz.Unmarshaler = (*StateRoots)(nil)
 
 // Byte32 represents a 32 bytes StateRoots object in Ethereum beacon chain consensus.
-type StateRoots [stateRootsSize][32]byte
+type StateRoots [8192][32]byte
```
Comment (rkapka) [inline beacon-chain/state/custom-types/state_arrays.go:66]: constant does not work with fastssz

File: beacon-chain/state/v1/setters_block.go [hunk 2]
```diff
@@ -52,14 +52,11 @@ func (b *BeaconState) UpdateBlockRootAtIndex(idx uint64, blockRoot [32]byte) err
 
 	r := b.state.BlockRoots
 	if ref := b.sharedFieldReferences[blockRoots]; ref.Refs() > 1 {
-		// Copy elements in underlying array by reference.
-		r = make([][]byte, len(b.state.BlockRoots))
-		copy(r, b.state.BlockRoots)
```
Comment (rkapka) [inline beacon-chain/state/v1/setters_block.go:57]: Please review that this is OK

File: beacon-chain/state/custom-types/state_arrays.go (hunk 1 above)
Comment (terencechain) [inline beacon-chain/state/custom-types/state_arrays.go:66]: would that work?

```suggestion
type StateRoots [params.BeaconConfig().SlotsPerHistoricalRoot][32]byte
```

**Review by terencechain — CHANGES_REQUESTED** [review]
Two questions:


1.)
Can `8192` be replaced with `SlotsPerHistoricalRoot` constant? This is going become a headache if we ever have to update any constant parameters

2.)
There's many patterns of the below:
```go
blockRoots := state.BlockRoots()
bRoots := make([][]byte, len(blockRoots))
for i := range bRoots {
	bRoots[i] = blockRoots[i][:]
}
```

Im thinking whether that's something we can do to improve upon readability

File: beacon-chain/core/transition/state.go [hunk 3]
```diff
@@ -86,21 +86,22 @@ func OptimizedGenesisBeaconState(genesisTime uint64, preState state.BeaconState,
...
+	zeroHash32 := params.BeaconConfig().ZeroHash
+	zeroHash := zeroHash32[:]
 
 	activeIndexRoots := make([][]byte, params.BeaconConfig().EpochsPerHistoricalVector)
 	for i := 0; i < len(activeIndexRoots); i++ {
 		activeIndexRoots[i] = zeroHash
 	}
 
-	blockRoots := make([][]byte, params.BeaconConfig().SlotsPerHistoricalRoot)
+	blockRoots := [8192][32]byte{}
```
[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/core/transition/state.go:97]: `params.BeaconConfig().SlotsPerHistoricalRoot` is not a constant expression. It does not compile.

File: beacon-chain/state/custom-types/state_arrays.go (hunk 1 above)
Comment (rkapka) [reply beacon-chain/state/custom-types/state_arrays.go:66]: Nope, not a constant expression.

**Review by terencechain — APPROVED** [review]
I think most of my feedbacks can be addressed on the feature branch level. This PR is good as it is to merge into feature branch and begin further experiments

File: beacon-chain/core/epoch/epoch_processing.go [hunk 4]
```diff
@@ -381,9 +381,19 @@ func ProcessHistoricalRootsUpdate(state state.BeaconState) (state.BeaconState, e
 	// Set historical root accumulator.
 	epochsPerHistoricalRoot := params.BeaconConfig().SlotsPerHistoricalRoot.DivSlot(params.BeaconConfig().SlotsPerEpoch)
 	if nextEpoch.Mod(uint64(epochsPerHistoricalRoot)) == 0 {
+		blockRoots := state.BlockRoots()
```
Comment (nisdas) [inline beacon-chain/core/epoch/epoch_processing.go:384]: This is going to be a big drag on performance, this whole piece of code does 4 copies

File: beacon-chain/state/v1/setters_state.go [hunk 5]
```diff
@@ -43,14 +43,11 @@ func (b *BeaconState) UpdateStateRootAtIndex(idx uint64, stateRoot [32]byte) err
 	// Check if we hold the only reference to the shared state roots slice.
 	r := b.state.StateRoots
 	if ref := b.sharedFieldReferences[stateRoots]; ref.Refs() > 1 {
-		// Copy elements in underlying array by reference.
-		r = make([][]byte, len(b.state.StateRoots))
```
Comment (nisdas) [inline beacon-chain/state/v1/setters_state.go:47]: same here

File: beacon-chain/state/v1/setters_block.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/state/v1/setters_block.go:57]: no it isn't, this will lead to a corrupted state removing this copy.

**Review by nisdas — COMMENTED** [review]
Just to clarify, fixed sized types are copied by value. So 2d-arrays just need an assignment to be copied:
https://play.golang.org/p/VfeuYERhV44

**Review by nisdas — APPROVED** [review]

**Pull request merged**